src/Engine/EmitMSIL/BuiltIn.cs

```diff
@@ -0,0 +1,383 @@
+        }
+
+        public static object mul(object a, object b)
+        {
+            if (a is int ai && b is int bi)
+            {
+                return HandleOverflow(() => checked(ai * bi), () => ai * bi);
+            }
+
+            if ((a is long || a is int) && (b is long || b is int))
+            {
+                var al = Convert.ToInt64(a);
+                var bl = Convert.ToInt64(b);
+                return HandleOverflow(() => checked(al * bl), () => al * bl);
+            }
+
+            if ((a is long || a is double || a is int) && (b is long || b is double || b is int))
+            {
+                return Convert.ToDouble(a) * Convert.ToDouble(b);
+            }
+
+            return null;
+        }
+
+        public static object div(object a, object b)
+        {
+            //division is always carried out as a double
+            if ((a is long || a is double || a is int) && (b is long || b is double || b is int))
+            {
+                return Convert.ToDouble(a) / Convert.ToDouble(b);
+            }
+
+            return null;
+        }
+
+        public static object mod(object a, object b)
+        {
+            if ((a is long || a is double || a is int) && (b is long || b is double || b is int))
+            {
+                if (a is int ri &&  b is int bi)
+                {
+                    var intMod = ri % bi;
+
+                    // In order to follow the conventions of Java, Python, Scala and Google's calculator,
+                    // the returned modulo will follow the sign of the divisor (not the dividend). 
+                    if (intMod < 0 && bi > 0 || intMod > 0 && bi < 0)
+                    {
+                        intMod = intMod + bi;
+                    }
+
+                    return intMod;
+                }
+
+                if ((a is long || a is int) && (b is long || b is int))
+                {
+                    var bl = Convert.ToInt64(b);
+                    var intMod = Convert.ToInt64(a) % bl;
+
+                    if (intMod < 0 && bl > 0 || intMod > 0 && bl < 0)
+                    {
+                        intMod = intMod + bl;
+                    }
+
+                    return intMod;
+                }
+
+                var bd = Convert.ToDouble(b);
+                var doubleMod = Convert.ToDouble(a) % bd;
+
+                if (doubleMod < 0 && bd > 0 || doubleMod > 0 && bd < 0)
+                {
+                    doubleMod = doubleMod + bd;
+                }
+
+                return doubleMod;
+            }
+
+            return null;
+        }
+
+        public static object Neg(object a)
+        {
+            switch (a)
+            {
+                case int ai:
+                    return HandleOverflow(() => checked(-ai), () => -ai);
+                case long al:
+                    return HandleOverflow(() => checked(-al), () => -al);
+                case double ad:
+                    return -ad;
+            }
```

Comment:
Can we write `if(a is long || a is int || a is double) return -a;`?

Reply:
No, does not work because `a` is an object.